Gem/Code/Source/Components/PerfTest/NetworkPrefabSpawnerComponent.cpp

```diff
@@ -0,0 +1,239 @@
+            }
+        }
+
+        return {};
+    }
+
+    void NetworkPrefabSpawnerComponent::CreateInstance(const SpawnRequest& request, const AssetItem* asset)
+    {
+        AZ_Assert(asset, "AssetMap didn't contain the asset id for prefab spawning");
+
+        AZ::Transform world = request.m_whereToSpawn;
+        if (asset)
+        {
+            auto ticket = AZStd::make_shared<AzFramework::EntitySpawnTicket>(asset->m_spawnableAsset);
+
+            auto preSpawnCallback = [world, request, ticket]([[maybe_unused]] AzFramework::EntitySpawnTicket::Id ticketId, AzFramework::SpawnableEntityContainerView view)
+            {
+                const AZ::Entity* rootEntity = *view.begin();
+                if (AzFramework::TransformComponent* entityTransform = rootEntity->FindComponent<AzFramework::TransformComponent>())
+                {
+                    entityTransform->SetWorldTM(world);
+                }
+
+                if (request.m_callbacks.m_beforeActivateCallback)
+                {
+                    request.m_callbacks.m_beforeActivateCallback(ticket, view);
+                }
+            };
+
+            auto onSpawnedCallback = [request, ticket]([[maybe_unused]] AzFramework::EntitySpawnTicket::Id ticketId, AzFramework::SpawnableConstEntityContainerView view)
+            {
+                if (request.m_callbacks.m_onActivateCallback)
+                {
+                    request.m_callbacks.m_onActivateCallback(ticket, view);
+                }
+            };
+
+            AZ_Assert(ticket->IsValid(), "Unable to instantiate spawnable asset");
+            if (ticket->IsValid())
+            {
+                AzFramework::SpawnAllEntitiesOptionalArgs optionalArgs;
+                optionalArgs.m_preInsertionCallback = AZStd::move(preSpawnCallback);
+                optionalArgs.m_completionCallback = AZStd::move(onSpawnedCallback);
+                AzFramework::SpawnableEntitiesInterface::Get()->SpawnAllEntities(*ticket, AZStd::move(optionalArgs));
+            }
+        }
+    }
+
+    void NetworkPrefabSpawnerComponent::OnAssetReady(AZ::Data::Asset<AZ::Data::AssetData> asset)
+    {
+        const AZ::Data::AssetId assetId = asset.GetId();
+        AZ::Data::AssetBus::MultiHandler::BusDisconnect(assetId);
+
+        const auto foundAsset = m_assetMap.find(assetId);
+        if (foundAsset != m_assetMap.end())
+        {
+            for (auto requestIterator = m_requests.begin(); requestIterator < m_requests.end(); /*iterating inside the loop body*/)
+            {
+                const SpawnRequest& request = *requestIterator;
+
+                if (request.m_assetIdToSpawn == assetId)
+                {
+                    CreateInstance(request, &foundAsset->second);
+                    requestIterator = m_requests.erase(requestIterator);
+                }
+                else
+                {
+                    ++requestIterator;
+                }
+            }
+        }
+    }
+}
```

Comment:
Would be good to route the player spawner through this, though there is a very surprising amount of code in this component.. And I kind of agree with Rajiv, there's a dizzying amount of new components here with unclear division of responsibilities between them. It would be good to clearly disambiguate what is a singleton system level component, and which are entity components that layer on top. Basically, I really hope our spawning abstractions are not so complex that customers would have to maintain a system style component like this one just to spawn network prefabs into the world.. that would indicate to me that our spawning API's have significantly missed the mark.